### PR TITLE
mds: add balance_automate fs setting

### PR DESCRIPTION
**Pull request by batrick**
To turn off automatic ("default") balancer in multiple MDS clusters. The new default is "off" as it is a constant source of problems and surprise administrators trying multiple actives. Instead, it should be a deliberate decision to turn it on and usually with customization like the "bal_rank_mask" setting.

Fixes: https://tracker.ceph.com/issues/61378

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
Thinking about a better name that `balance_automate`. Maybe `balancer_automation`. We have a mix of `verb_object` and `object` style names so there isn't any way to be consistent. We do have other balance related settings beginning with `bal` so this seemed reasonable.

**Comment by batrick** [comment]
Still thinking about what tests to include here. Maybe a switch in `fs:workload` to set random ephemeral pins or turn on the balancer.

**Comment by lxbsz** [comment]
> Thinking about a better name that `balance_automate`. Maybe `balancer_automation`. We have a mix of `verb_object` and `object` style names so there isn't any way to be consistent. We do have other balance related settings beginning with `bal` so this seemed reasonable.

How about `automate_balancer` ?

**Comment by vshankar** [comment]
So, pre upgrade if some subtrees were migrated to other ranks, those subtrees continue to be handled by the respective ranks post upgrade? Or, would it be the case that reducing max_mds to one (1) during upgrade will rebalance all subtrees to rank 0 where the balancer is disabled (post upgrade)?

**Comment by batrick** [comment]
> So, pre upgrade if some subtrees were migrated to other ranks, those subtrees continue to be handled by the respective ranks post upgrade?

Yes, unless balancing is re-enabled or pinning is used.

> Or, would it be the case that reducing max_mds to one (1) during upgrade will rebalance all subtrees to rank 0 where the balancer is disabled (post upgrade)?

That could happen too, depending how they upgrade. There's the fail_fs option for cephadm some folks may be using.

**Comment by batrick** [comment]
Keeping `balance_automate` for now. This is ready for review.

cc @vshankar

**Comment by vshankar** [comment]
jenkins retest this please

**Review by lxbsz — APPROVED** [review]
LGTM.

**Comment by vshankar** [comment]
@batrick You mentioned in standup that you found a bug while testing this. Are you running this through fs suite yourself? If yes, then I can drop this from my branch.

**Comment by batrick** [comment]
I'm running fs:workload to confirm the qa tests work. Once done, a broader qa run would be appropriate.

**Comment by vshankar** [comment]
> (So, yes, go ahead and drop it from your set.)

ACK. Tag me when its ready for a broader integration run.

**Comment by batrick** [comment]
Test run: https://pulpito.ceph.com/pdonnell-2023-07-21_01:10:44-fs:workload-wip-pdonnell-testing-20230720.223327-distro-default-smithi/

If nothing unexpected then this is ready for general QA.

**Comment by batrick** [comment]
Lots of failed counter changes due to pinning. I'll look into a fix next week.

**Comment by batrick** [comment]
jenkins test windows

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2023-08-23_03:59:53-fs-wip-vshankar-testing-20230822.060131-testing-default-smithi/

(will review when the run finishes)

File: qa/suites/fs/workload/ranks/multi/balancer/distributed.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+tasks:
+- exec:
+    mon.a:
+      - ceph fs set cephfs balance_automate false
+      - ceph fs subvolumegroup pin cephfs qa distributed 1
```
Comment (vshankar) [inline qa/suites/fs/workload/ranks/multi/balancer/distributed.yaml:5]: There are a handful of test failures with the following error

> The following counters failed to be set on mds daemons: {'mds.imported', 'mds.exported'}

See: https://pulpito.ceph.com/vshankar-2023-08-23_03:59:53-fs-wip-vshankar-testing-20230822.060131-testing-default-smithi/7377124/

It reaches to the point where `MDBalancer::handle_export_pins` is selecting an mds rank for each of the dirfrags (the split is done way before hitting the omap limit which is expected with epins),but, I think the selected rank is not a peer mds and therefore the migrator does not get involved.

https://github.com/ceph/ceph/blob/main/src/mds/MDBalancer.cc#L210

From mds log: ./remote/smithi194/log/3ae5e002-416c-11ee-9b3a-001a4aab830c/ceph-mds.i.log.gz

```
2023-08-23T04:48:33.468+0000 7f805f947700 20 mds.0.bal handle_export_pins export_pin_queue size=1
2023-08-23T04:48:33.468+0000 7f805f947700 20 mds.0.bal handle_export_pins  executing export_pin=-2 on [inode 0x10000000001 [...22,head] /volumes/qa/ auth{1=1,2=1} fragtree_t(*^4) v1254 DIRTYPARENT DISTEPHEMERALPIN f(v0 m2023-08-23T04:27:31.735295+0000 2=0+2) n(v55 rc2023-08-23T04:47:40.459278+0000 b9054571302 1077=1043+34) old_inodes=28 (inest mix) (iversion lock) | dirtyscattered=0 request=0 lock=0 dirfrag=16 caps=0 dirtyrstat=0 dirtyparent=1 replicated=1 dirty=1 waiter=0 authpin=0 scrubqueue=0 distepin 0x55c9cc38ab00]
2023-08-23T04:48:33.468+0000 7f805f947700 15 mds.0.bal handle_export_pins number of auth trees = 10; not printing auth trees
```

Maybe the call to

> mdcache->hash_into_rank_bucket(cd->ino(), cd->get_frag());

is always returning the current mds rank - not sure yet.

**Comment by batrick** [comment]
> Lots of failed counter changes due to pinning. I'll look into a fix next week.

tests in 4 hours: https://shaman.ceph.com/builds/ceph/wip-batrick-testing-20230831.182458/

**Comment by vshankar** [comment]
> > Lots of failed counter changes due to pinning. I'll look into a fix next week.
> 
> tests in 4 hours: https://shaman.ceph.com/builds/ceph/wip-batrick-testing-20230901.180327

Is this pending review of the test run @batrick ?

**Comment by vshankar** [comment]
> > > Lots of failed counter changes due to pinning. I'll look into a fix next week.
> > 
> > 
> > tests in 4 hours: https://shaman.ceph.com/builds/ceph/wip-batrick-testing-20230901.180327
> 
> Is this pending review of the test run @batrick ?

@batrick ping?

**Comment by batrick** [comment]
> > > > Lots of failed counter changes due to pinning. I'll look into a fix next week.
> > > 
> > > 
> > > tests in 4 hours: https://shaman.ceph.com/builds/ceph/wip-batrick-testing-20230901.180327
> > 
> > 
> > Is this pending review of the test run @batrick ?
> 
> @batrick ping?

I think that test run didn't start because of build failures. I will revisit it.

**Comment by batrick** [comment]
In four hours: https://pulpito.ceph.com/?branch=wip-batrick-testing-20231019.005909

**Comment by batrick** [comment]
(trivial rebase)

**Comment by batrick** [comment]
jenkins test dashboard

**Comment by batrick** [comment]
jenkins test dashboard cephadm

**Comment by batrick** [comment]
jenkins test dashboard cephadm

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcOeXUIG191898-wip-yuri2-testing-2023-12-06-1239-old-wip-yuri2-testing-2023-12-04-0902

**Comment by ljflores** [comment]
jenkins test api

**Comment by yuriw** [comment]
@batrick this was tested
pls merge
ref: https://trello.com/c/OeXUIG19

**Comment by yuriw** [comment]
jenkins test api

**Comment by batrick** [comment]
trivial rebase

**Comment by batrick** [comment]
jenkins test dashboard cephadm

**Comment by batrick** [comment]
Added docs.

**Comment by batrick** [comment]
```
The following tests FAILED:
	236 - unittest_rbd_mirror (SEGFAULT)
```

https://tracker.ceph.com/issues/63798

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test dashboard

**Comment by batrick** [comment]
jenkins test dashboard cephadm

**Comment by batrick** [comment]
jenkins test dashboard cephadm

**Comment by vshankar** [comment]
jenkins test dashboard cephadm

**Comment by batrick** [comment]
jenkins test dashboard cephadm

**Comment by zdover23** [comment]
@batrick, should the "Dynamic Subtree Partitioning" section of the docs in `doc/cephfs/multimds.rst` be included in the Quincy release branch?

**Comment by batrick** [comment]
> @batrick, should the "Dynamic Subtree Partitioning" section of the docs in `doc/cephfs/multimds.rst` be included in the Quincy release branch?

No, this setting is not present in quincy.

**Pull request merged**